src/clev2er/utils/cs2/retrackers/gaussian_exponential_retracker.py

```diff
@@ -0,0 +1,88 @@
+"""Module for the gaussian plus exponential retracker (or Giles' retracker)
+
+gauss_plus_exp_tracker(): the retracking function which fits a gaussian plus exponential function
+    to the input waveform to find the retracking point.
+_GaussPlusExp(): the gaussian exponential function used by gauss_plus_exp_tracker
+
+Author: Ben Palmer
+Date: 01 Mar 2024
+"""
+
+import warnings
+
+import numpy as np
+import numpy.typing as npt
+from scipy.optimize import curve_fit
+
+
+def gauss_plus_exp_tracker(waveform: npt.NDArray, max_iterations: int = 3000) -> float:
+    """Gaussian plus exponential retracker, or Giles' retracker.
+
+    Retracks waveforms using a gaussian plus exponential function. Uses Levenberg-Marquardt
+    non-linear least-squares method to fit the function to the waveform. Returns the location of the
+    maximum amplitude of the fitted wave as the retracking point.
+
+    Args:
+        waveform (npt.NDArray): Input waveform
+
+    Raises:
+        RuntimeError: Raised if GaussPlusExp function encounters an error while fitting.
+
+    Returns:
+        float: The retracking point
+    """
+
+    warnings.simplefilter("ignore", RuntimeWarning, 84)
+
+    x: npt.NDArray = np.arange(waveform.size).astype(float)
+
+    x0: np.intp = np.argmax(waveform)
+    a: float = waveform[x0]
+
+    try:
+        popt, _ = curve_fit(  # pylint: disable=unbalanced-tuple-unpacking
+            _gauss_plus_exp, x, waveform[:], p0=[a, x0, 1, 1], maxfev=max_iterations, method="lm"
+        )
+    except RuntimeError:
+        return np.nan
+
+    if "popt" in locals():
+        tracking_point = popt[1]
```

Comment:
Possibly this section should be restructured as tracking_point will be undefined if "popt" not in locals()

Suggest something like
tracking_point = np.nan  # Initialize tracking_point with a default value

    try:
        popt, _ = curve_fit(
            _gauss_plus_exp, x, waveform, p0=[a, x0, 1, 1], maxfev=max_iterations, method="lm"
        )
        tracking_point = popt[1]  # Update tracking_point if fitting is successful
    except RuntimeError:
        pass  # tracking_point remains np.nan if fitting fails

    return tracking_point